File: src/main/ipc/feeds.js
// getHotTagFromData helper function used by the hot search handler
function getHotTagFromData(item) {
  const showName = item.show_name || ''
  if (showName.includes('新') || showName.includes('回归')) return '新'
  if (showName.includes('独家')) return '独家'
  if (showName.includes('番') || showName.includes('动画')) return 'bangumi'
  if (showName.includes('视频') || showName.includes('直播')) return 'video'
  return ''
}

function registerFeedsHandlers(deps) {
  const { ipcMain, fetchWithRetry, buildRecommendUrl, fetchApi, log, fetchWbiKeys, getMixKey, signParams } = deps

  ipcMain.handle('test-ipc', async () => {
    console.log('Test IPC called')
    return { success: true, message: 'IPC works!', data: [1, 2, 3] }
  })

  ipcMain.handle('fetch-videos', async (event, page = 1) => {
    log('fetch-videos called, page:', page)
    try {
      const url = buildRecommendUrl(page)
      log('Using recommend API:', url)
      const result = await fetchWithRetry(url)
      if (result.success && result.data.code === 0) {
        log('Recommend API success, code:', result.data.code)
        return { success: true, data: result.data }
      }
      return { success: false, error: '获取推荐视频失败' }
    } catch (error) {
      log('Recommend API failed:', error.message)
      return { success: false, error: error.message }
    }
  })

  ipcMain.handle('search-videos', async (event, keyword, page = 1, searchType = 'all', order = 'totalrank') => {
    log('search-videos called, keyword:', keyword, 'page:', page, 'type:', searchType, 'order:', order)

    try {
      // 根据搜索类型确定 page_size、search_type、ad_resource
      const typeConfig = {
        'all':            { pageSize: 42, searchType: null,     adResource: '5646' },
        'video':          { pageSize: 42, searchType: 'video',  adResource: '5654' },
        'media_bangumi':  { pageSize: 12, searchType: 'media_bangumi', adResource: '5646' },
        'media_ft':       { pageSize: 12, searchType: 'media_ft',      adResource: '5646' },
        'bili_user':      { pageSize: 36, searchType: 'bili_user',     adResource: '5646' }
      }
      const config = typeConfig[searchType] || typeConfig['all']

      // 构建请求参数（参与 WBI 签名），对齐真实 Bilibili 搜索接口参数
      const params = {
        keyword: keyword,
        page: page,
        page_size: config.pageSize,
        platform: 'pc',
        highlight: 1,
        single_column: 0,
        from_source: 'web_search',
        from_spmid: '333.337',
        source_tag: 3,
        web_location: '1430654',
        ad_resource: config.adResource,
        __refresh__: 'true',
        _extra: '',
        context: '',
        pubtime_begin_s: 0,
        pubtime_end_s: 0,
        category_id: '',
        gaia_vtoken: ''
      }

      if (config.searchType) {
        params.search_type = config.searchType
      }

      // 综合搜索特有参数
      if (searchType === 'all') {
        params.duration = ''
        params.web_roll_page = 1
      }

      // 视频搜索支持排序参数
      if (searchType === 'video') {
        params.dynamic_offset = 0
        params.web_roll_page = 1
      }

      // 视频和综合搜索始终发送 order 参数（综合排序为空字符串，与 B站接口对齐）
      if (searchType === 'all' || searchType === 'video') {
        params.order = (order && order !== 'totalrank') ? order : ''
      }

      // 番剧搜索
      if (searchType === 'media_bangumi') {
        params.duration = ''
        params.order = ''
      }

      // 影视搜索
      if (searchType === 'media_ft') {
        params.duration = ''
        params.order = ''
      }

      // 用户搜索需要额外参数
      if (searchType === 'bili_user') {
        params.order_sort = 0
        params.user_type = 0
        params.dynamic_offset = 0
      }

      // WBI 签名
      const keys = await fetchWbiKeys()
      if (!keys || !keys.imgKey) {
        log('WBI keys 获取失败，尝试不带签名请求')
        // 不带签名直接请求作为后备
        const baseUrl = config.searchType
          ? 'https://api.bilibili.com/x/web-interface/wbi/search/type'
          : 'https://api.bilibili.com/x/web-interface/wbi/search/all/v2'
        const queryString = Object.keys(params)
          .map(k => `${encodeURIComponent(k)}=${encodeURIComponent(String(params[k]))}`)
          .join('&')
        const endpoint = `${baseUrl}?${queryString}`
        const result = await fetchWithRetry(endpoint)
        if (result.success && result.data.code === 0) {
          return { success: true, data: result.data }
        }
        return { success: false, error: '搜索失败' }
      }

      const mixKey = getMixKey(keys.imgKey, keys.subKey)
      const signed = signParams(params, mixKey)

      // 拼接最终 URL（签名参数 w_rid 和 wts 附加到末尾）
      const allParams = { ...params, w_rid: signed.w_rid, wts: signed.wts }
      const queryString = Object.keys(allParams)
        .map(k => `${encodeURIComponent(k)}=${encodeURIComponent(String(allParams[k]))}`)
        .join('&')

      const baseUrl = config.searchType
        ? 'https://api.bilibili.com/x/web-interface/wbi/search/type'
        : 'https://api.bilibili.com/x/web-interface/wbi/search/all/v2'
      const endpoint = `${baseUrl}?${queryString}`

      log('Using search endpoint:', endpoint.substring(0, 120) + '...')
      const result = await fetchWithRetry(endpoint)
      if (result.success && result.data.code === 0) {
        log('Search API成功, code:', result.data.code)
        return { success: true, data: result.data }
      }

      log('Search API失败, code:', result.data?.code, 'message:', result.data?.message)
      return { success: false, error: result.data?.message || '搜索失败' }
    } catch (error) {
      log('Search API错误:', error.message)
      return { success: false, error: error.message }
    }
  })

  ipcMain.handle('fetch-popular-videos', async (event, ...args) => {
    // 兼容旧版本调用方式（只传 page）
    let tab = 'comprehensive'
    let page = 1
    let rid = 0

    if (args.length === 1 && typeof args[0] === 'number') {
      // 旧版本调用：只传 page
      page = args[0]
      log('fetch-popular-videos called (legacy mode), page:', page)
    } else {
      // 新版本调用
      tab = args[0] || 'comprehensive'
      page = args[1] || 1
      rid = args[2] || 0
      log('fetch-popular-videos called, tab:', tab, 'page:', page, 'rid:', rid)
    }

    try {
      let result = null
      let endpoint = ''

      if (tab === 'comprehensive' || typeof tab === 'number') {
        // 综合热门接口
        endpoint = `https://api.bilibili.com/x/web-interface/popular?ps=40&pn=${page}&web_location=bilibili-electron`
        log('Using popular endpoint:', endpoint)
        result = await fetchWithRetry(endpoint)
      } else if (tab === 'ranking') {
        // 排行榜接口
        endpoint = `https://api.bilibili.com/x/web-interface/ranking/v2?rid=${rid}&type=all&ps=30&pn=${page}&web_location=bilibili-electron`
        log('Using ranking/v2 endpoint:', endpoint)
        result = await fetchWithRetry(endpoint)
      } else if (tab === 'weekly') {
        // 先获取最新期数
        let latestNumber = 375
        try {
          const listEndpoint = 'https://api.bilibili.com/x/web-interface/popular/series/list'
          const listResult = await fetchWithRetry(listEndpoint)
          if (listResult && listResult.success && listResult.data?.code === 0) {
            const seriesList = listResult.data?.data?.list || []
            if (seriesList.length > 0) {
              latestNumber = seriesList[0].number
            }
          }
        } catch (e) {
          log('获取每周必看期数列表失败，使用默认值:', e.message)
        }

        // 每周必看接口需要 WBI 签名
        const params = { number: latestNumber, web_location: 'bilibili-electron' }
        const keys = await fetchWbiKeys()
        const mixKey = getMixKey(keys.imgKey, keys.subKey)
        const signed = signParams(params, mixKey)
        endpoint = `https://api.bilibili.com/x/web-interface/popular/series/one?number=${latestNumber}&web_location=bilibili-electron&w_rid=${signed.w_rid}&wts=${signed.wts}`
        log('Using weekly endpoint:', endpoint)
        result = await fetchWithRetry(endpoint)
      } else if (tab === 'precious') {
        endpoint = `https://api.bilibili.com/x/web-interface/popular/precious`
        log('Using precious endpoint:', endpoint)
        result = await fetchWithRetry(endpoint)
      }

      if (result && result.success) {
        log('API成功, raw data:', JSON.stringify(result.data).substring(0, 500))
        return { success: true, data: result.data }
      }
      log('API失败, result:', result)
      return { success: false, error: '获取视频失败' }
    } catch (error) {
      log('fetch-popular-videos error:', error.message)
      return { success: false, error: error.message }
    }
  })

  ipcMain.handle('fetch-popular-videos-v2', async (event, page = 1) => {
    log('fetch-popular-videos-v2 called, page:', page)
    try {
      const endpoint = `https://api.bilibili.com/x/web-interface/ranking/v2?rid=0&type=all&ps=30&pn=${page}`
      log('Using ranking/v2 endpoint:', endpoint)
      const result = await fetchApi(endpoint)
      log('Popular API result code:', result.code)
      if (result.code === 0) {
        log('Popular API success, items:', result.data?.list?.length || 0)
        return { success: true, data: result }
      }
      log('Popular API failed:', result)
      return { success: false, data: result, error: result.message || '获取热门视频失败' }
    } catch (error) {
      log('fetch-popular-videos-v2 error:', error.message)
      return { success: false, error: error.message }
    }
  })

  ipcMain.handle('fetch-hot-search', async (event) => {
    log('fetch-hot-search called')
    try {
      const endpoint = 'https://api.bilibili.com/x/web-interface/wbi/search/square?limit=10&platform=web&web_location=333.1365&w_rid=33c27013429cc439349b6d7f3523bbb8&wts=1777972362'
      log('Using hot search endpoint:', endpoint)
      const result = await fetchWithRetry(endpoint)

      if (result && result.success && result.data) {
        const apiData = result.data
        if (apiData.code === 0 && apiData.data && apiData.data.trending && apiData.data.trending.list) {
          const hotList = apiData.data.trending.list.map(item => ({
            keyword: item.keyword || item.show_name || '',
            title: item.show_name || item.keyword || '',
            tag: getHotTagFromData(item)
          }))
          log('Hot search API成功, items count:', hotList.length)
          return { success: true, data: { list: hotList } }
        } else if (apiData.trending && apiData.trending.list) {
          const hotList = apiData.trending.list.map(item => ({
            keyword: item.keyword || item.show_name || '',
            title: item.show_name || item.keyword || '',
            tag: getHotTagFromData(item)
          }))
          log('Hot search API成功(备用格式), items count:', hotList.length)
          return { success: true, data: { list: hotList } }
        }
      }

      log('热搜API返回数据格式不正确')
      return { success: false, error: '数据格式不正确' }

    } catch (error) {
      log('Hot search API错误:', error.message)
      return { success: false, error: error.message }
    }
  })
}

module.exports = { registerFeedsHandlers }

File: src-tauri/src/ipc/feeds.rs
// feeds 模块：对应 Electron 版 src/main/ipc/feeds.js
// 通道：test-ipc / fetch-videos / search-videos / fetch-popular-videos /
//       fetch-popular-videos-v2 / fetch-hot-search
use serde_json::{json, Value};

use crate::api::{
    build_recommend_url, fetch_api, fetch_with_retry, fetch_wbi_keys, get_mix_key, sign_params,
    wrap_err, wrap_ok,
};

fn args_u64(args: &[Value], idx: usize, default: u64) -> u64 {
    args.get(idx)
        .and_then(|v| v.as_u64().or_else(|| v.as_f64().map(|f| f as u64)))
        .unwrap_or(default)
}

fn args_str(args: &[Value], idx: usize, default: &str) -> String {
    args.get(idx)
        .and_then(|v| v.as_str())
        .unwrap_or(default)
        .to_string()
}

/// test-ipc
pub async fn test_ipc() -> Value {
    json!({ "success": true, "message": "IPC works!", "data": [1, 2, 3] })
}

/// fetch-videos：首页推荐
pub async fn fetch_videos(args: &[Value]) -> Value {
    let page = args_u64(args, 0, 1);
    let url = build_recommend_url(page);
    match fetch_with_retry(&url).await {
        Ok(data) => {
            let code = data.get("code").and_then(|c| c.as_i64()).unwrap_or(-1);
            if code == 0 {
                wrap_ok(data)
            } else {
                wrap_err("获取推荐视频失败")
            }
        }
        Err(e) => wrap_err(&e),
    }
}

/// search-videos：全站搜索（WBI 签名）
pub async fn search_videos(args: &[Value]) -> Value {
    let keyword = args_str(args, 0, "");
    let page = args_u64(args, 1, 1);
    let search_type = args_str(args, 2, "all");
    let order = args_str(args, 3, "totalrank");

    // 根据搜索类型确定 page_size、search_type、ad_resource
    let (page_size, st, ad_resource): (u64, Option<&str>, &str) = match search_type.as_str() {
        "video" => (42, Some("video"), "5654"),
        "media_bangumi" => (12, Some("media_bangumi"), "5646"),
        "media_ft" => (12, Some("media_ft"), "5646"),
        "bili_user" => (36, Some("bili_user"), "5646"),
        _ => (42, None, "5646"),
    };

    let mut params: Vec<(String, String)> = vec![
        ("keyword".into(), keyword),
        ("page".into(), page.to_string()),
        ("page_size".into(), page_size.to_string()),
        ("platform".into(), "pc".into()),
        ("highlight".into(), "1".into()),
        ("single_column".into(), "0".into()),
        ("from_source".into(), "web_search".into()),
        ("from_spmid".into(), "333.337".into()),
        ("source_tag".into(), "3".into()),
        ("web_location".into(), "1430654".into()),
        ("ad_resource".into(), ad_resource.into()),
        ("__refresh__".into(), "true".into()),
        ("_extra".into(), "".into()),
        ("context".into(), "".into()),
        ("pubtime_begin_s".into(), "0".into()),
        ("pubtime_end_s".into(), "0".into()),
        ("category_id".into(), "".into()),
        ("gaia_vtoken".into(), "".into()),
    ];

    if let Some(st) = st {
        params.push(("search_type".into(), st.to_string()));
    }
    match search_type.as_str() {
        "all" => {
            params.push(("duration".into(), "".into()));
            params.push(("web_roll_page".into(), "1".into()));
            params.push((
                "order".into(),
                if order == "totalrank" { "".into() } else { order.clone() },
            ));
        }
        "video" => {
            params.push(("dynamic_offset".into(), "0".into()));
            params.push(("web_roll_page".into(), "1".into()));
            params.push((
                "order".into(),
                if order == "totalrank" { "".into() } else { order.clone() },
            ));
        }
        "media_bangumi" | "media_ft" => {
            params.push(("duration".into(), "".into()));
            params.push(("order".into(), "".into()));
        }
        "bili_user" => {
            params.push(("order_sort".into(), "0".into()));
            params.push(("user_type".into(), "0".into()));
            params.push(("dynamic_offset".into(), "0".into()));
        }
        _ => {}
    }

    let base_url = if st.is_some() {
        "https://api.bilibili.com/x/web-interface/wbi/search/type"
    } else {
        "https://api.bilibili.com/x/web-interface/wbi/search/all/v2"
    };

    let build_query = |params: &[(String, String)]| -> String {
        params
            .iter()
            .map(|(k, v)| format!("{}={}", crate::api::enc(k), crate::api::enc(v)))
            .collect::<Vec<_>>()
            .join("&")
    };

    // WBI 签名
    match fetch_wbi_keys().await {
        Some((img_key, sub_key)) => {
            let mix_key = get_mix_key(&img_key, &sub_key);
            let (w_rid, wts) = sign_params(&params, &mix_key);
            let mut all = params.clone();
            all.push(("w_rid".into(), w_rid));
            all.push(("wts".into(), wts.to_string()));
            let endpoint = format!("{}?{}", base_url, build_query(&all));
            match fetch_with_retry(&endpoint).await {
                Ok(data) => {
                    let code = data.get("code").and_then(|c| c.as_i64()).unwrap_or(-1);
                    if code == 0 {
                        wrap_ok(data)
                    } else {
                        wrap_err(
                            data.get("message")
                                .and_then(|m| m.as_str())
                                .unwrap_or("搜索失败"),
                        )
                    }
                }
                Err(e) => wrap_err(&e),
            }
        }
        None => {
            // WBI keys 获取失败：不带签名直接请求作为后备
            let endpoint = format!("{}?{}", base_url, build_query(&params));
            match fetch_with_retry(&endpoint).await {
                Ok(data) => {
                    let code = data.get("code").and_then(|c| c.as_i64()).unwrap_or(-1);
                    if code == 0 {
                        wrap_ok(data)
                    } else {
                        wrap_err("搜索失败")
                    }
                }
                Err(e) => wrap_err(&e),
            }
        }
    }
}

/// fetch-popular-videos：热门（comprehensive/ranking/weekly/precious 四种 tab）
pub async fn fetch_popular(args: &[Value]) -> Value {
    // 兼容旧版本调用方式（只传 page 数字）
    let legacy = args.len() == 1 && args[0].is_number();
    let tab = if legacy {
        "comprehensive".to_string()
    } else {
        args_str(args, 0, "comprehensive")
    };
    let page = if legacy {
        args_u64(args, 0, 1)
    } else {
        args_u64(args, 1, 1)
    };
    let rid = if legacy { 0 } else { args_u64(args, 2, 0) };

    let endpoint = match tab.as_str() {
        "ranking" => format!(
            "https://api.bilibili.com/x/web-interface/ranking/v2?rid={}&type=all&ps=30&pn={}&web_location=bilibili-electron",
            rid, page
        ),
        "weekly" => {
            // 先获取最新期数
            let mut latest_number: u64 = 375;
            if let Ok(list_data) =
                fetch_with_retry("https://api.bilibili.com/x/web-interface/popular/series/list")
                    .await
            {
                if let Some(number) = list_data
                    .pointer("/data/list/0/number")
                    .and_then(|n| n.as_u64())
                {
                    latest_number = number;
                }
            }
            // 每周必看需要 WBI 签名
            let Some((img_key, sub_key)) = fetch_wbi_keys().await else {
                return wrap_err("WBI keys 获取失败");
            };
            let mix_key = get_mix_key(&img_key, &sub_key);
            let params = vec![
                ("number".to_string(), latest_number.to_string()),
                ("web_location".to_string(), "bilibili-electron".to_string()),
            ];
            let (w_rid, wts) = sign_params(&params, &mix_key);
            format!(
                "https://api.bilibili.com/x/web-interface/popular/series/one?number={}&web_location=bilibili-electron&w_rid={}&wts={}",
                latest_number, w_rid, wts
            )
        }
        "precious" => "https://api.bilibili.com/x/web-interface/popular/precious".to_string(),
        _ => format!(
            "https://api.bilibili.com/x/web-interface/popular?ps=40&pn={}&web_location=bilibili-electron",
            page
        ),
    };

    match fetch_with_retry(&endpoint).await {
        Ok(data) => wrap_ok(data),
        Err(e) => wrap_err(&e),
    }
}

/// fetch-popular-videos-v2：排行榜
pub async fn fetch_popular_v2(args: &[Value]) -> Value {
    let page = args_u64(args, 0, 1);
    let endpoint = format!(
        "https://api.bilibili.com/x/web-interface/ranking/v2?rid=0&type=all&ps=30&pn={}",
        page
    );
    match fetch_api(&endpoint).await {
        Ok(data) => {
            let code = data.get("code").and_then(|c| c.as_i64()).unwrap_or(-1);
            if code == 0 {
                wrap_ok(data)
            } else {
                json!({
                    "success": false,
                    "data": data,
                    "error": data.get("message").and_then(|m| m.as_str()).unwrap_or("获取热门视频失败")
                })
            }
        }
        Err(e) => wrap_err(&e),
    }
}

fn hot_tag_from_data(item: &Value) -> String {
    let show_name = item
        .get("show_name")
        .and_then(|s| s.as_str())
        .unwrap_or("");
    if show_name.contains('新') || show_name.contains("回归") {
        return "新".into();
    }
    if show_name.contains("独家") {
        return "独家".into();
    }
    if show_name.contains('番') || show_name.contains("动画") {
        return "bangumi".into();
    }
    if show_name.contains("视频") || show_name.contains("直播") {
        return "video".into();
    }
    String::new()
}

fn map_hot_list(list: &[Value]) -> Vec<Value> {
    list.iter()
        .map(|item| {
            let keyword = item
                .get("keyword")
                .and_then(|k| k.as_str())
                .or_else(|| item.get("show_name").and_then(|k| k.as_str()))
                .unwrap_or("");
            let title = item
                .get("show_name")
                .and_then(|k| k.as_str())
                .or_else(|| item.get("keyword").and_then(|k| k.as_str()))
                .unwrap_or("");
            json!({ "keyword": keyword, "title": title, "tag": hot_tag_from_data(item) })
        })
        .collect()
}

/// fetch-hot-search：热搜榜
pub async fn fetch_hot_search(_args: &[Value]) -> Value {
    // 与 Electron 版一致：使用固定签名的 square 接口
    let endpoint = "https://api.bilibili.com/x/web-interface/wbi/search/square?limit=10&platform=web&web_location=333.1365&w_rid=33c27013429cc439349b6d7f3523bbb8&wts=1777972362";
    match fetch_with_retry(endpoint).await {
        Ok(api_data) => {
            let code = api_data.get("code").and_then(|c| c.as_i64()).unwrap_or(-1);
            if code == 0 {
                if let Some(list) = api_data.pointer("/data/trending/list").and_then(|l| l.as_array()) {
                    return wrap_ok(json!({ "list": map_hot_list(list) }));
                }
            }
            if let Some(list) = api_data.pointer("/trending/list").and_then(|l| l.as_array()) {
                return wrap_ok(json!({ "list": map_hot_list(list) }));
            }
            wrap_err("数据格式不正确")
        }
        Err(e) => wrap_err(&e),
    }
}
